File: app/src/main/java/com/demo/justapp/exchanger/data/repository/RatesRepositoryImpl.java
package com.demo.justapp.exchanger.data.repository;

import android.support.annotation.NonNull;

import com.demo.justapp.exchanger.data.converter.RatesConverter;
import com.demo.justapp.exchanger.data.network.RestApi;
import com.demo.justapp.exchanger.domain.repository.RatesRepository;

import io.reactivex.Single;

/**
 * Реализация {@link RatesRepository}
 *
 * @author Sergey Rodionov
 */
public class RatesRepositoryImpl implements RatesRepository {

    private final RestApi mRestApi;
    private final RatesConverter mRatesConverter;

    /**
     * Конструктор для {@link RatesRepositoryImpl}
     *
     * @param restApi        API для работы с сервером
     * @param ratesConverter конвертор информации о курсах валют в локальную модель
     */
    public RatesRepositoryImpl(@NonNull RestApi restApi,
                               @NonNull RatesConverter ratesConverter) {
        mRestApi = restApi;
        mRatesConverter = ratesConverter;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Single<CurrencyRatesModel> loadRates(@NonNull String rate) {
        return mRestApi.getRates(rate)
                .map(mRatesConverter::convert);
    }
}

File: app/src/main/java/com/demo/justapp/exchanger/data/repository/RatesRepositoryImpl.kt
package com.demo.justapp.exchanger.data.repository

import com.demo.justapp.exchanger.data.converter.RatesConverter
import com.demo.justapp.exchanger.data.network.RestApi
import com.demo.justapp.exchanger.di.scope.Data
import com.demo.justapp.exchanger.domain.repository.RatesRepository
import com.demo.justapp.exchanger.models.local.CurrencyRatesModel
import com.demo.justapp.exchanger.models.remote.RatesResponse
import io.reactivex.Single
import javax.inject.Inject

/**
 * Реализация [RatesRepository]
 *
 * @author Sergey Rodionov
 */
@Data
class RatesRepositoryImpl @Inject constructor(restApi: RestApi,
                                              ratesConverter: RatesConverter) : RatesRepository {

    private val mRestApi: RestApi = checkNotNull(restApi)
    private val mRatesConverter: RatesConverter = checkNotNull(ratesConverter)

    override fun loadRates(rate: String): Single<CurrencyRatesModel> {
        return mRestApi.getRates(rate).map { ratesResponse: RatesResponse ->
            mRatesConverter.convert(ratesResponse)
        }
    }

}